peer/gossip.go

```diff
@@ -0,0 +1,148 @@
+package peer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/renproject/aw/dht"
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+	"go.uber.org/zap"
+)
+
+type gossipFunc func(ctx context.Context, subnet id.Hash, contentID, content []byte) error
+
+// Gossiper returns a new set of callback functions and a gossip function that can be used to spread information throughout a network
+
+func Gossiper(t *transport.Transport, contentResolver dht.ContentResolver, addressTable Table, logger zap.Logger, next Callbacks) (Callbacks, gossipFunc) {
+
+	gossip := func(ctx context.Context, subnet id.Hash, contentID, content []byte) error {
+		msg := wire.Msg{Version: wire.MsgVersion1, Type: wire.MsgTypePush, Data: contentID}
+
+		var chainedError error = nil
+		for _, sig := range addressTable.Subnet() {
+			addr, ok := addressTable.PeerAddress(sig)
+			if !ok {
+				logger.Error("gossip", zap.String("table", "peer not found"))
+				continue
+			}
+			if err := t.Send(ctx, sig, addr, msg); err != nil {
+				if chainedError == nil {
+					chainedError = fmt.Errorf("%v, gossiping to peer %v : %v", chainedError, sig, err)
+				} else {
+					chainedError = fmt.Errorf("gossiping to peer %v : %v", sig, err)
+				}
+			}
+		}
+		return chainedError
+	}
+
+	didReceiveSync := func(from id.Signatory, msg wire.Msg) {
+	}
+	didReceiveSyncAck := func(from id.Signatory, msg wire.Msg) {
+	}
+	didReceivePush := func(from id.Signatory, msg wire.Msg) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		response := wire.Msg{
+			Type: wire.MsgTypePushAck,
+			Data: nil,
+		}
+
+		addr, ok := addressTable.PeerAddress(from)
+		if !ok {
+			logger.Error("gossip", zap.String("table", "peer not found"))
+			return
+		}
```

Comment:
If it gets to the callback with the id of a remote peer, the address should exist in the peer table, otherwise we have no way of sending them a pull request